### PR TITLE
Fix regexp_replace using null

### LINKED ISSUE #44274: regexp_replace fails with null replacement string
In the field calculator, one can use `regexp_replace(input, regex, replacement)` to conditionally replace part of a string. If the replacement is Null, however, the output will always be Null even if there is no text matching the regex expression

`regexp_replace('aaaa','b',null) --> Null`

Expected output: 
```
regexp_replace('aaaa','b',null) --> aaaa (no change)
regexp_replace('aaaa','a',null) --> Null (some text is found and replaced by Null, so the all output is Null)
```

It is a problem when the replacement value is stored in a field and contains Null entries that shouldn't be considered
Ex:
```  
FieldA FieldB
---------------------
aaa     b
zzz     NULL
aaa     c

fieldC = regexp_replace(fieldA,'a',fieldB)

FieldA FieldB       fieldC-current   fieldC-expected
--------------------------------------------
aaa     b            bbb             bbb
zzz     NULL         NULL            zzz
aaa     c            ccc             ccc
```

Workaround:
coalesce first...   `regexp_replace(fieldA,'b',coalesce(fieldB,''))`

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

The expression regexp_replace does not handle null value and produce unwanted results when a field has one.

fixes #44274

### CODE REVIEW

File: src/core/expression/qgsexpressionfunction.cpp [hunk 1]
```diff
@@ -1485,6 +1485,8 @@ static QVariant fcnRegexpReplace( const QVariantList &values, const QgsExpressio
   QString str = QgsExpressionUtils::getStringValue( values.at( 0 ), parent );
   QString regexp = QgsExpressionUtils::getStringValue( values.at( 1 ), parent );
   QString after = QgsExpressionUtils::getStringValue( values.at( 2 ), parent );
+  if ( after.isNull() )
+    after = QString( "" );
```
Comment (nyalldawson) [inline src/core/expression/qgsexpressionfunction.cpp:1489]: Is this change actually required? I think the change to the function definition alone should be sufficient to fix the issue. Null QStrings are mostly a historical oddity...

File: tests/src/core/testqgsexpression.cpp [hunk 2]
```diff
@@ -1403,6 +1403,8 @@ class TestQgsExpression: public QObject
       QTest::newRow( "regexp_replace non greedy" ) << "regexp_replace('HeLLo','(?<=H).*?L', '-')" << false << QVariant( "H-Lo" );
       QTest::newRow( "regexp_replace cap group" ) << "regexp_replace('HeLLo','(eL)', 'x\\\\1x')" << false << QVariant( "HxeLxLo" );
       QTest::newRow( "regexp_replace invalid" ) << "regexp_replace('HeLLo','[[[', '-')" << true << QVariant();
+      QTest::newRow( "regexp_replace match, with null" ) << "regexp_replace('aaaa','a',NULL)" << false << QVariant( "" );
```
Comment (nyalldawson) [inline tests/src/core/testqgsexpression.cpp:1406]:
```suggestion
      QTest::newRow( "regexp_replace match, with null" ) << "regexp_replace('aaaa','a',NULL)" << false << QVariant();
```

File: src/core/expression/qgsexpressionfunction.cpp (hunk 1 above)
Comment (lbartoletti) [reply src/core/expression/qgsexpressionfunction.cpp]: Nope sorry, old and bad habit

File: tests/src/core/testqgsexpression.cpp (hunk 2 above)
Comment (lbartoletti) [reply tests/src/core/testqgsexpression.cpp:1406]: Nope, it should return an empty string

Comment (nyalldawson) [reply tests/src/core/testqgsexpression.cpp:1406]: That seems counter intuitive to me. Wouldn't you need to use an empty string as the replacement character to get that? (How does postgres behave in this situation)

File: tests/src/core/testqgsexpression.cpp [hunk 3]
```diff
@@ -1403,6 +1403,8 @@ class TestQgsExpression: public QObject
...
+      QTest::newRow( "regexp_replace no match with null" ) << "regexp_replace('zzz', 'b', NULL)" << false << QVariant( "zzz" ); // fix https://github.com/qgis/QGIS/issues/44274
```
Comment (m-kuhn) [inline tests/src/core/testqgsexpression.cpp:1407]: Can you add one for `regexp_replace('aaabbb', 'b', NULL)` as well?

File: tests/src/core/testqgsexpression.cpp (hunk 2 above)
Comment (lbartoletti) [reply tests/src/core/testqgsexpression.cpp:1406]: @nyalldawson some tests on postgresql:

`SELECT regexp_replace('aaa', 'a', NULL) is NULL` => True
`SELECT regexp_replace('aaa', 'b', NULL) is NULL` => True
SELECT regexp_replace('abaa', 'a', NULL) is NULL => True
SELECT regexp_replace('abaa', 'b', NULL) is NULL => True
SELECT regexp_replace('abaa', 'c', NULL) is NULL => True

With our expression engine and Qt replace : `NULL` is considered as a `QVariant(Invalid)` and `str.replace` with `QString()` or `QString('')` will return the same string:

`QString("aaa").replace(QString("a"), QString())` => QString("")
`QString("aaabbb").replace(QString("b"), QString())` => QString("aaa")

So for you, if a NULL is sent as third parameter, we must return a `QVariant()`?

File: tests/src/core/testqgsexpression.cpp (hunk 3 above)
Comment (lbartoletti) [reply tests/src/core/testqgsexpression.cpp:1408]: done.

Comment (m-kuhn) [reply tests/src/core/testqgsexpression.cpp:1408]: Thanks, I'd say this should also be NULL (`QVariant()`), just like `"" + NULL ==> NULL`

File: src/core/expression/qgsexpressionfunction.cpp [hunk 4]
```diff
@@ -1482,6 +1482,10 @@ static QVariant fcnReplace( const QVariantList &values, const QgsExpressionConte
 
 static QVariant fcnRegexpReplace( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
 {
+  if ( ! values.at( 2 ).isValid() )
```
Comment (m-kuhn) [inline src/core/expression/qgsexpressionfunction.cpp:1485]:
```suggestion
  if ( ! values.at( 2 ).isNull() )
```

Better safe than sorry, an invalid is also null

**Comment by m-kuhn** [comment]
Thanks, code looks good to me.
One final request, can you document this behavior in the function doc?

File: tests/src/core/testqgsexpression.cpp (hunk 2 above)
Comment (nyalldawson) [reply tests/src/core/testqgsexpression.cpp]: I think if NULL is 3rd parameter AND the search string is found then yes, it should return null instead of empty string. If search string is NOT found then we should return original string.

File: src/core/expression/qgsexpressionfunction.cpp [hunk 5]
```diff
@@ -1485,13 +1485,20 @@ static QVariant fcnRegexpReplace( const QVariantList &values, const QgsExpressio
   QString str = QgsExpressionUtils::getStringValue( values.at( 0 ), parent );
   QString regexp = QgsExpressionUtils::getStringValue( values.at( 1 ), parent );
   QString after = QgsExpressionUtils::getStringValue( values.at( 2 ), parent );
+  bool isValid_after{ values.at( 2 ).isValid() };
```
Comment (m-kuhn) [inline src/core/expression/qgsexpressionfunction.cpp:1488]: Can you make that `isNul()`?
It will work with invalid and (valid) Null QVariants..

**Comment by roya0045** [comment]
Was there a plan to resurrect this? This is why I'm not that fond of the stalebot for PR...
